content/articles/custom-decorators-in-django/index.md

```diff
@@ -0,0 +1,499 @@
+    <p>{{ message }}</p>
+{% endfor %}
+<h1>Profile</h1>
+```
+templates/accounts/register.html
+```html
+<form action="{% url 'accounts:register' %}" method="post">
+    {% csrf_token %}
+    {{ form.as_p }}
+    <input type="submit" value="Register">
+</form>
+```
+The login.html file is the same as the register.html, just replace register with login.
+Create a new file called `urls.py` to map the route / path to the right view function
+```py
+from django.urls import path
+from .views import (login, profile, register)
+
+app_name = 'accounts'
+urlpatterns = [
+    path('login', login, name="login"),
+    path('register', register, name="register"),
+    path('profile', profile, name="profile"),
+]
+```
+Testing the application written so far…
+Create a superuser and then start your server
+```bash
+python manage.py createsuperuser
+python manage.py runserver
+```
+Go to your browser and navigate to http://127.0.0.1:8000/
+
+Check the image below
+![Authentication Decorators in Action](custom-decorators-in-django/authdec.gif)
+
+
+- A link or an OTP should be sent to the user's email address in order to verify his / her email account and activate the user. I have kept this tutorial simple that the admin should make the user active.
+> The default authentication backend does not return a user object for in active users. So if a user who created an account using the register.html won't be able to login unless the admin makes the user active.
+
+Now to implement the xhr_request_only decorator created earlier.
+Inside posts/views.py
+```py
+from django.http import JsonResponse
+from django.shortcuts import get_object_or_404, render
+from .decorators import xhr_request_only
+from accounts.decorators import verification_required
+from django.contrib.auth.decorators import login_required
+from django.views.decorators.http import require_POST
+from .models import Post
+import json
+
+def post_detail_view(request, id):
+    post = get_object_or_404(Post, id=id)
+    return render(request, 'posts/detail.html', {'post': post})
+
+@verification_required
+@login_required(login_url="accounts:login")
+@xhr_request_only
+@require_POST
+def post_vote_view(request):
+    data  = json.loads(request.body.decode("utf-8"))
+    id = data.get('postId')
+    action = data.get('action')
+    post = get_object_or_404(Post, id=id)
+    if action == "upvote":
+        post.downvote_users.remove(request.user)
+        post.upvote_users.add(request.user)
+    elif action == "downvote":
+        post.upvote_users.remove(request.user)
+        post.downvote_users.add(request.user)
+    return JsonResponse({"message": action})
+```
+In the above code,
+
+- Line 3 is to import xhr_request_only from the decorators module.
+- Line 4 is to import verification_required from the decorators module in the accounts application.
+- The built-in require_POST decorator has been used on the post_vote_view to ensure  that it only accepts the POST request method.
```

Comment:
Replace with...
The built-in require_POST decorator has been used on the 'post_vote_view' to ensure that it only accepts the 'POST' request method.